### PR TITLE
Add visited_without_bp / missed_visits to Reporting SchemaV2

### PR DESCRIPTION
**Pull request by rsanheim**
* grab these numbers from schema v2 and handle the delegation
* rename the _rate method to _rates to be consistent with everything
else
* ~~add a calculated sum which we need to get visited_without_bp under
care AND LTFU~~ this was incorrect and not necessary

**Story card:** [ch4515](https://app.clubhouse.io/simpledotorg/story/4515/grab-missed-visits-and-visited-w-o-bp-from-reporting-pipeline)

### CODE REVIEW

File: app/assets/javascripts/common/reports.js
Comment (rsanheim) [inline app/assets/javascripts/common/reports.js:665]: Getting names more consistent here.

File: app/controllers/reports/regions_controller.rb
Comment (rsanheim) [inline app/controllers/reports/regions_controller.rb:77]: Realized these calls weren't passing thru the v2 flag - fixing here.

File: app/models/reports/repository.rb
Comment (rsanheim) [inline app/models/reports/repository.rb:175]: none of this is needed anymore

File: app/models/reports/schema_v2.rb
Comment (rsanheim) [inline app/models/reports/schema_v2.rb:223]: This helped catch some things in dev/test with a slightly better error.

File: spec/models/reports/repository_spec.rb
Comment (rsanheim) [inline spec/models/reports/repository_spec.rb:354]: This was the best I got to to make this specs readable while still covering a wide range of periods and data.

File: spec/models/reports/repository_spec.rb
Comment (rsanheim) [inline spec/models/reports/repository_spec.rb:466]: Just accounting for buggy behavior in v1 code here.

**Review by rsanheim — COMMENTED** [review]
Ready for review - leaving some notes inline for context

File: app/models/reports/schema_v2.rb
Comment (ssrihari) [inline app/models/reports/schema_v2.rb:131]: `:missed_visits_rates` seems to be an alias for 2 methods?

File: app/models/reports/schema_v2.rb
Comment (ssrihari) [inline app/models/reports/schema_v2.rb:143]: Hey, I think the definition here and other ltfu/non-ltfu queries may be incorrect. The `_lost_to_follow_up` numbers in the facility states table has the total for **only** lost to follow up patients, not ltfu+under_care, which is probably what we want here. Do our tests cover this well?

So we should be either summing up the under_care and ltfu numbers here, or exposing ltfu+under_care numbers from the table. In previous discussions about this, I suggested exposing ltfu, and under_care numbers separately so that we can use them as building blocks here in ruby land. Exposing ltfu+under care numbers from the DB would be quite rigid in terms of design.

File: app/models/reports/schema_v2.rb
Comment (ssrihari) [inline app/models/reports/schema_v2.rb:150]: Same question [as in previous PR](https://github.com/simpledotorg/simple-server/pull/2767/files#r684976196): Is this is for convenience, or do we need caching here?

File: app/models/reports/schema_v2.rb
Comment (ssrihari) [reply app/models/reports/schema_v2.rb:143]: Also, `missed_visits` [defined above](https://github.com/simpledotorg/simple-server/pull/2839/files#diff-e95110f3347ec2a8767b4146482eff9bf0685c7ae55302c630070cbe3411084bR116) does `field = with_ltfu ?`.. it would be good to be consistent in structuring these methods.

File: config/initializers/cache_stats.rb
Comment (ssrihari) [inline config/initializers/cache_stats.rb:3]: Was this a debugging change?

**Review by ssrihari — CHANGES_REQUESTED** [review]
Only one blocking change around the definition of the ltfu numbers.

File: app/models/reports/schema_v2.rb
Comment (rsanheim) [reply app/models/reports/schema_v2.rb:131]: Yeah, I wanted to move to a plural `rates` method names, but `missed_visit_rate` is currently used in too many places to change, hence the alias here so that I can start changing it incrementally in another PR.

File: config/initializers/cache_stats.rb
Comment (rsanheim) [reply config/initializers/cache_stats.rb:3]: It gets too noisy to have this enabled all the time in dev, but it is very useful for cache debugging and monitoring in dev.  I'll had some better comments or tooling around this to make that more clear.

File: app/models/reports/schema_v2.rb
Comment (rsanheim) [reply app/models/reports/schema_v2.rb:143]: I actually had the summing code here and then removed it after working thru tests and reconsidering if we needed it.  Now with fresh eyes I think you are right that we need it.  If we do need it, I think we should add it to the facility_states matview to keep the app side code straightforward...the additional logic to sum both fields got a little ugly and also made everything just inconsistent enough in SchemaV2 to be frustrating.

I'm going to reasses this and will open a PR to add the summed columns if need be.

Comment (rsanheim) [reply app/models/reports/schema_v2.rb]: @ssrihari Okay, I handled this in code now and updated all the specs. This also gets the numbers much more in line when viewing things in reports, so I think this is good to go after an approval.

We can add it to the matviews next week if we want, though the code to add it was not too bad after all.

File: app/models/reports/schema_v2.rb
Comment (rsanheim) [reply app/models/reports/schema_v2.rb:155]: I think we will want some caching, when I compare this to v1 code its slightly slower for very large production like data sets.  But I think we can revisit performance a bit later (see https://app.clubhouse.io/simpledotorg/story/4516/benchmark-reports-comparing-v1-to-v2).

**Comment by rsanheim** [comment]
@ssrihari This is ready for another pass.

File: app/models/reports/schema_v2.rb
Comment (kpethtel) [inline app/models/reports/schema_v2.rb:26]: Could throw on a freeze here for consistency.

File: spec/models/reports/repository_spec.rb
Comment (kpethtel) [inline spec/models/reports/repository_spec.rb:472]: These changes to testing the schema were the one thing that threw me here. It seems like this approach to testing would have worked before but maybe wasn't a critical distinction until now.

**Review by kpethtel — APPROVED** [review]
This is difficult to review as someone who doesn't regularly work in this area of code, but I don't see any issues so 🚢

File: app/models/reports/schema_v2.rb
Comment (harimohanraj89) [inline app/models/reports/schema_v2.rb:238]: Ah I see, this is the crux of handling the LTFU toggle. Nice.

**Review by harimohanraj89 — APPROVED** [review]
I'm gonna lean a little bit on you to double-check the test coverage, but the big moving parts look great. :shipit:

File: app/models/reports/schema_v2.rb
Comment (rsanheim) [reply app/models/reports/schema_v2.rb:238]: 👍

File: spec/models/reports/repository_spec.rb
Comment (rsanheim) [reply spec/models/reports/repository_spec.rb:472]: yeah these cache tests will probably go away soon, so I'm not too worried about them.

**Pull request merged**
